File: ch09/random_select.py

import random


def partition(v, left, right):
  i = left - 1
  for j in range(left, right):
    if v[j] < v[right]:
      i += 1
      exchange(v, i, j)
  i += 1
  exchange(v, i, right)
  return i


def random_partition(v, left, right):
  p = random.randint(left, right)
  exchange(v, p, right)
  return partition(v, left, right)


def random_select_b(index, v, left, right):
  def rs_help(l, r):
    if l < r:
      p = random_partition(v, l, r)
      if index < p:
        rs_help(l, p - 1)
      elif index > p:
        rs_help(p + 1, r)
      #else:
      return v[index]
    else:
      return v[index]

  return rs_help(left, right)


def random_select(index, v, left, right):
  if left < 0:
    raise ValueError("left")
  elif right >= len(v):
    raise ValueError("right")
  elif index < left or index > right:
    raise ValueError("index")
  else:
    return random_select_b(index, v, left, right)


#---------------------------------------

def insertion_sort(v):
  n = len(v)
  if n > 0:
    for i in range(1, n):
      p = i - 1
      q = i
      while p >= 0 and v[p] > v[q]:
        exchange(v, p, q)
        p -= 1
        q -= 1


def exchange(v, i, j):
  temp = v[i]
  v[i] = v[j]
  v[j] = temp


def mkRandomList(size, n_min, n_max):
  v = []
  for i in range(0, size):
    v.append(random.randint(n_min, n_max - 1))
  return v


def printList(v):
  print('(List: #', end='')
  print(len(v), end='')
  print(' [', end='')
  for i in range(0, len(v)):
    print(v[i], end='')
    print(' ', end='')
  print('])')


A = mkRandomList(100, 0, 100)
printList(A)
print(random_select(5, A, 0, 99))
insertion_sort(A)
printList(A)

File: ch09/random_select.js

const partition = function(v, left, right) {
    let i = left - 1;
    for (let j = left; j != right; j += 1) {
        if (v[j] < v[right]) {
            i += 1;
            exchange(v, i, j);
        }
    }
    i += 1;
    exchange(v, i, right);
    return i;
}

const random_partition = function(v, left, right) {
    let p = randomInt(left, right + 1);
    exchange(v, p, right);
    return partition(v, left, right);
}

const random_select_b = function(index, v, left, right) {
    let rs_help = function(l, r) {
        if (l < r) {
            let p = random_partition(v, l, r);
            if (index < p) {
                rs_help(l, p - 1);
            } else if (index > p) {
                rs_help(p + 1, r);
            } //else {
                return v[index];
            //}
        } else {
            return v[index];
        }
    }

    return rs_help(left, right);
}

const random_select = function(index, v, left, right) {
    if (left < 0) {
        throw new IllegalArgumentError("left");
    } else if (right >= v.length) {
        throw new IllegalArgumentError("right");
    } else if (index < left || index > right) {
        throw new IllegalArgumentError("index");
    } else {
        return random_select_b(index, v, left, right);
    }
}

//--------------------------------------

class IllegalArgumentError extends Error {
  constructor(message) {
    super(message);
    this.message = message;
    this.name = 'IllegalArgumentError';
  }
}

const insertion_sort = function(v) {
    let n = v.length;
    if (n > 0) {
        for (let i = 1; i != n; i += 1) {
            for (let p = i - 1, q = i;
                     p >= 0 && v[p] > v[q]; p -= 1, q -= 1) {
                exchange(v, p, q);
            }
        }
    }
}

const exchange = function(v, i, j) {
    let temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

const randomInt = function(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
}

const mkIntArray = function(size, min, max) {
    let v = new Array();
    for (let i = 0; i < size; i += 1) {
        v.push(randomInt(min, max));
    }
    return v;
}

const printArray = function(v) {
    process.stdout.write("(Array #" + v.length + " [")
    for (element of v) {
        process.stdout.write(element + " ");
    }
    process.stdout.write("])\n");
}

let A = mkIntArray(100, 0, 100);
printArray(A);
console.log(random_select(95, A, 0, 99));
insertion_sort(A);
printArray(A);
